Source/DafnyCore/ProofDependencyWarnings.cs

```diff
@@ -86,6 +86,15 @@ public class ProofDependencyWarnings {
         return false;
       }
 
+      // Don't warn about `assert false` being proved vacuously. If it's proved,
+      // it must be vacuous, but it's also probably an attempt to prove that a
+      // given alternative is unreachable.
+      var assertedExpr = poDep.ProofObligation.GetAssertedExpr(options);
+      if (assertedExpr is not null &&
+          Expression.IsBoolLiteral(assertedExpr, out var lit) &&
+          lit == false) {
```

Comment:
Although I think this is a good point, I'm somewhat inclined to leave the behavior as-is. The reason is that, for _any_ proof, it may be that there are intermediate assertions that are, at one point, necessary for the proof and then later become unnecessary because of verifier changes. I think it's less surprising to users if proving something by contradiction behaves the same way as proving something directly. It's a less than optimal in both cases that the set of things that were needed to complete a proof may change over time. But I think it's better if it changes the same way for both direct and by-contradiction proofs.

Reply:
ok, then it's a separate problem we can think of in the long term.